### PR TITLE
Refactor demask

### LINKED ISSUE #293: Refactoring of `DeMask` model
### Goal
Hey guys, I was planning to refactor `DeMask` to use `BaseEncoderMaskerDecoder` so that it supports 1D input data when traced with 2D (and the other way round). Currently, a traced `DeMask` model will only work with data that has the same input shape as the example used for tracing because of the way `forward` is written.

Refactoring the class with `BaseEncoderMaskerDecoder` would automatically solve this issue in addition to cleaning the class. However I had some questions before making a whole bunch of changes (see below).

#### in `__init__`

1. When computing `n_feats_input` with `fb_type` set to `stft`, isn't the value 
```python
n_feats_input = (self.encoder.filterbank.n_filters) // 2 + 1
```
only valid for even values of `n_filters`. If it is, wouldn't we need to either support the case of odd values for `n_filters` or raise an error when `n_filters` is odd?
2. When computing `n_feats_output`, there is a check on `self.input_type`. Shouldn't it be on `self.output_type` instead? The error message also mentions "Input type" but shouldn't it be about "Output type"?

#### In `forward`

The current flow of data in the `forward` of `BaseEncoderMaskerDecoder` is as follow (written as plain text as Mermaid diagrams are not supported on GitHub yet):

```
wav -> <input shaping> -> [Encoder] -> [Encoder postprocess] -> [Encoder activation] -> tf_rep

tf_rep ->  [Masker] -> masker_outmasker_out --> [Masker postprocess] -> <mult> -> masked 
       |                                                                  1
       |------------------------------------------------------------------|
masked -> [Decoder] -> [Decoder postprocess] -> <shape output> -> reconstruction
```
where the values in square brackets indicate blocks that can be changed by any inheriting class and the values in `<>` indicate operations that are hard coded in the forward of `BaseEncoderMaskerDecoder`.

As the forward of `DeMask` is written, the input to the mask estimator is not always `tf_rep`. To make it compatible with the `BaseEncoderMaskerDecoder`, we could add a `preprocess_masker_input` between `tf_rep` and `[Masker]` in the base class. Like all the hooks in `BaseEncoderMaskerDecoder`, this would have a default behaviour of returning the data given as input.

### PR DESCRIPTION
**Pull request by mhu-coder**
Fixes #293 

Quick summary of the changes:

- [ | ] Backward compatibility tests for `DeMask`. The tests are commented at the moment while waiting for a solution to #294 
- [x] Clean up of `__init__`
- [x] Fix masker I/O dimensions when either `input_type` or `output_type` are set to `reim`
- [x] Changes to `BaseEncoderMaskerDecoder.forward` to cover what is going on inside `DeMask`:
  - [x] Add `preprocess_masker_input` hook
  - [x] Add `preprocess_product_input` hook, which processes `tf_rep` before application of `masks`
  - [x] Fix `_shape_reconstructed`: we should have tested the length of `size` instead of checking its `ndim` attribute as `ndim` of a `tensor.shape` is always 1
- [x] Make `DeMask` inherit from `BaseEncoderMaskerDecoder`
  - [x] Remove `forward`
  - [x] Rewrite `__init__`
  - [x] Override `preprocess_masker_input`, `preprocess_product_input` and `postprocess_masks`
- [   ] Check that recipe still runs
- [x] Raise error for `STFT` filter bank with odd number of filters.

:hankey: `preprocess_product_input` does not seem right, but I did not find a nicer way of making `DeMask` compatible with the forward of `BaseEncoderMaskerDecoder`. Any suggestion welcome :)!
:hankey: `DeMask` takes a `fb_type` argument while all the other model classes (and `make_enc_dec`) use a `fb_name` argument. Changing it could break backward compatibility though :/

### CODE REVIEW

**Comment by mpariente** [comment]
Thanks a lot for the PR !

I'm going to leave the first review round to @JorisCos on this PR.
Just answering this point here. 

>  `DeMask` takes a `fb_type` argument while all the other model classes (and `make_enc_dec`) use a `fb_name` argument. Changing it could break backward compatibility though :/

I know... It was merged in a hurry, never again ! :wink: 
We're going to minorly break BC in 0.4.0. We'll also update the Zenodo records for 0.4.0, so we can change to `fb_name` at that point and the pretrained model will be loadable on both 0.3.4 and 0.4.0, I think it's reasonnable.

File: asteroid/models/base_models.py [hunk 1]
```diff
@@ -305,10 +305,12 @@ def forward(self, wav):
         tf_rep = self.postprocess_encoded(tf_rep)
         tf_rep = self.enc_activation(tf_rep)
 
-        est_masks = self.masker(tf_rep)
```
Comment (jonashaag) [inline asteroid/models/base_models.py:308]: To me all of this looks like a special case that ought not to handle in this generic forward() method. We don’t have to make forward() fit to every possible use case, otherwise it will become so generic that the abstraction hasn’t any meaning anymore. 

How about simply moving this entire mask application code into a method `apply_mask(tf_rep, est_masks) -> masked_tf_rep`? Or we can simply not use the parent forward() method at all.

Comment (mpariente) [reply asteroid/models/base_models.py:308]: I was thinking about `apply_mask` too. This is better. 

I think `preprocess_masker_input` makes sense. If the filterbank is the STFT and you want some phase features, you don't want to mask them. It's pretty common IMO. Alternatively, we could have a `forward_masker` or `apply_masker` function instead of applying the masker directly, and this could contain additional feature extraction in subclasses.

Comment (jonashaag) [reply asteroid/models/base_models.py:308]: Maybe it would have been better to not implement these hooks like
that in the first place and only split into 3 methods `forward_{endoder,masker,decoder}` from the beginning... not sure

Comment (mhu-coder) [reply asteroid/models/base_models.py:308]: > Or we can simply not use the parent forward() method at all.

The incentive to stick with the `forward` was to not copy paste the block of code that shapes the input/output.

> split into 3 methods `forward_{endoder,masker,decoder}`

I agree. Having a `forward_{encoder, masker, decoder}` offers more flexibility to the inheriting classes. In addition to these 3 blocks, I would add an `apply_mask` method to keep the mask application step separate from the previous blocks. Along these lines, the forward of `BaseEncoderMaskerDecoder` would look like:

```python
def forward(self, wav):
   shape = jitable_shape(wav)
   wav = _unsqueeze_to_3d(wav)

   tf_rep = self.forward_encoder(wav)
   est_masks = self.forward_masker(tf_rep)
   masked_tf_rep = self.apply_mask(tf_rep, est_masks)
   decode = self.forward_decoder(masked_tf_rep)

   reconstructed = pad_x_to_y(decoded, wav)
   return _shape_reconstructed(reconstructed, shape)

def forward_encoder(self, wav):
    # Computes time-frequency representation of `wav`
    # default: applies encoder and encoder activation
    # ...
    return tf_rep

def forward_masker(self, tf_rep):
    # Estimate masks based on `tf_rep`.
    # default: applies masker
    # ...
    return masks

def apply_masks(self, tf_rep, masks):
    # Applies `masks` on `tf_rep`
    return masks * tf_rep.unsqueeze(1)

def forward_decoder(self, masked):
    # Decodes masked time-frequency representation
    # default: applies decoder
    return decoded
```

With this structure, the only thing that is imposed to any class that inherits from `BaseEncoderMaskerDecoder` is the following diagram (again, in plain text):

```
(wav) --> Encoder --(tf_rep)----------------------> Mask_application ---(masked)---> Decoder ---> (decoded)
                            |                               1
                            |--> Mask_estimation --(masks)--|
```

Pre/post-processing would then be written directly in the inheriting classes.

For backward compatibility with classes that currently implements the `postprocess_*` hooks, we could call these in the `forward_*` and completely remove the hooks for v0.4? We would need to have tests ensuring backward compatibility for each of these models though :stuck_out_tongue:

Comment (mpariente) [reply asteroid/models/base_models.py:308]: Also not sure but this does look good, and maybe more flexible. I'll need a bit of time to think about it. 

Apart from `DCUNet` and `DCCRN`, no model uses that IIRC, and we have no pretrained models for them, plus `DCCRN` is going to change.

Comment (jonashaag) [reply asteroid/models/base_models.py:308]: Reminds me of this :) 
convtasnet
(from the Conv TasNet paper)

Comment (mpariente) [reply asteroid/models/base_models.py:308]: I don't know this, what is ConvTasNet?

Comment (JorisCos) [reply asteroid/models/base_models.py:308]: Hello ! I agree from what i have seen in `DCUNet` and `DCCRN` the changes are small and straightforward. Having `forward_{endoder,masker,decoder}` seems like the good way to go :)

**Comment by mpariente** [comment]
So, let's make this

```python
def forward(self, wav):
   shape = jitable_shape(wav)
   wav = _unsqueeze_to_3d(wav)

   tf_rep = self.forward_encoder(wav)
   est_masks = self.forward_masker(tf_rep)
   masked_tf_rep = self.apply_mask(tf_rep, est_masks)
   decode = self.forward_decoder(masked_tf_rep)

   reconstructed = pad_x_to_y(decoded, wav)
   return _shape_reconstructed(reconstructed, shape)
```

And call the current hooks in those four methods. 

Who'd like to do it? @mhu-coder ?
We'll block this PR until then.

**Comment by mhu-coder** [comment]
Sure. Quick questions, do I remove the backward compatibility tests since these
* are commented for CI
* would be added in #294 ?

**Comment by mpariente** [comment]
Cool !
Let's leave this PR as is for now, and open a new PR to change `BaseEncoderMaskerDecoder`. Once the other one is merged, we'll come back here.

**Comment by jonashaag** [comment]
Hook part LGTM

File: asteroid/models/demask.py [hunk 2]
```diff
@@ -66,98 +87,59 @@ def __init__(
...
 
-        self.encoder, self.decoder = make_enc_dec(
-            fb_type,
-            kernel_size=kernel_size,
-            n_filters=n_filters,
-            stride=stride,
-            sample_rate=sample_rate,
-            **fb_kwargs,
-        )
+    def _get_n_feats_input(self, input_type, encoder):
```
Comment (mpariente) [inline asteroid/models/demask.py:90]: Take `n_feats_out` as input instead of encoder?

File: asteroid/models/demask.py [hunk 3]
```diff
@@ -66,98 +87,59 @@ def __init__(
...
-        if wav.ndim == 2:
-            wav = wav.unsqueeze(1)
-        # Real forward
-        tf_rep = self.encoder(wav)
+        n_feats_input = encoder.n_feats_out // 2
+        if input_type == "cat":
+            n_feats_input += encoder.n_feats_out
+        return n_feats_input
+
+    def _get_n_feats_output(self, output_type, encoder):
```
Comment (mpariente) [inline asteroid/models/demask.py:102]: Same here.

File: asteroid/models/demask.py [hunk 4]
```diff
@@ -178,3 +160,31 @@ def get_model_args(self):
         }
         model_args.update(self.fb_kwargs)
         return model_args
+
+
+def _build_masker_nn(
```
Comment (mpariente) [inline asteroid/models/demask.py:165]: I wonder if this is the good way to go. 
I know the architecture is really simple, but should this be a hidden function? 
For other models, we have a `nn.Module` that can be used with other features than asteroid filterbanks. 
This might be related to @popcornell 's comments on asteroid's source code organization.

File: asteroid/models/demask.py [hunk 5]
```diff
@@ -178,3 +160,31 @@ def get_model_args(self):
...
+                nn.Conv1d(in_chan, hidden_dim, 1),
+                make_layer_norm(hidden_dim),
+                layer_activation,
+                nn.Dropout(dropout),
+            ]
+        )
+        in_chan = hidden_dim
+
+    net.extend([nn.Conv1d(in_chan, n_out, 1), activations.get(mask_act)()])
+    return net
```
Comment (mpariente) [inline asteroid/models/demask.py:190]: And why return the list and not the `Sequential` of it?

File: asteroid/models/demask.py [hunk 6]
```diff
@@ -178,3 +160,31 @@ def get_model_args(self):
         }
         model_args.update(self.fb_kwargs)
         return model_args
+
```
Comment (mpariente) [inline asteroid/models/demask.py:163]: If we're refactoring we should probably use `self.encoder.get_config()` in `get_model_args`. 
If we make a `DeMaskMasker` module (bad name), with `get_config`, we don't need to write `get_model_args`. This is related to the next comment.

**Review by mpariente — COMMENTED** [review]
This looks good in general. 
It raises some questions on code organization: where should masker architectures live? 

The current way is to have them in `masknn` and extend them to have encoder/decoder in `models`, is this the right way to do it? 
@popcornell suggests that masker architectures should be moved into `models` and `masknn` renamed to `nnet` or similar. 
What are your takes on this @mhu-coder @jonashaag ?

Even if demask's architecture is very simple, the current code is not consistent with the rest of the codebase and I think we should stay consistent.

File: asteroid/models/demask.py (hunk 4 above)
Comment (popcornell) [reply asteroid/models/demask.py:165]: maybe having it as a method would be better. 

> we have a nn.Module that can be used with other features than asteroid filterbanks.
I must have missed this. What is this module ?

Comment (mpariente) [reply asteroid/models/demask.py:165]: > maybe having it as a method would be better.

A method in the class? Or a function outside (like now)? public or private? 

> I must have missed this. What is this module ?

Sorry, it was not clear. I mean the masker architectures e.g `TDConvNet` can be used on anything and is exposed to the user. Here, the masker architecture from `DeMask` is not.

**Comment by mhu-coder** [comment]
> Even if demask's architecture is very simple, the current code is not consistent with the rest of the codebase and I think we should stay consistent.

Agree, it should stay consistent with the rest of the code base.

> @popcornell suggests that masker architectures should be moved into models and masknn renamed to nnet or similar.
What are your takes on this @mhu-coder @jonashaag ?

Looking at the base code, it seems that maskers are never used on their own and only in 1 model. So it would make sense to have them close to the models they define.
OTOH, having them in `asteroid.models` also makes it harder to compose them, if this need ever rises. I guess, for me it comes down to 2 questions:
1. How do you make the distinction between a "masker" architecture (which would go in `asteroid.models`) and a "building-block" architecture that can be used in a "masker" architecture (which would go in `asteroid.nnet` presumably)?
2. What is the need to move them at this point?

File: asteroid/models/demask.py (hunk 4 above)
Comment (popcornell) [reply asteroid/models/demask.py:165]: > A method in the class? Or a function outside (like now)? public or private?

private Method in the class. 

> Sorry, it was not clear. I mean the masker architectures e.g TDConvNet can be used on anything and is exposed to the user. Here, the masker architecture from DeMask is not.

OK, well it is a simple MLP the masker architecture.... i don't know how much is useful for other tasks. 

BTW, can't in principle `BaseENcoderMaskerDecoder` be used ? Or it can't because of difference for STFT mag vs free for example.

**Comment by popcornell** [comment]
> 1. How do you make the distinction between a "masker" architecture (which would go in asteroid.models) and a "building-block" architecture that can be used in a "masker" architecture (which would go in asteroid.nnet presumably)?
> 
> 2. What is the need to move them at this point?

I would say the distinction is dictated by how much that collection of layers is tied to one particular paper/architecture. 
Like the masker from convtasnet is quite binded to the whole convtasnet. Instead the DilateConvs blocks are also used in other architectures for example. 

No need to move for now, but as the models grow it can get confusing maybe.

File: asteroid/models/demask.py (hunk 4 above)
Comment (mpariente) [reply asteroid/models/demask.py:165]: This PR makes that we use `BaseEncoderMaskerDecoder` for `DeMask` now. 

> OK, well it is a simple MLP the masker architecture.... i don't know how much is useful for other tasks.

I know that it's a very simple architecture but I wonder if it makes a difference. 

Too many choices for today for me :joy:

File: asteroid/models/demask.py (hunk 6 above)
Comment (mhu-coder) [reply asteroid/models/demask.py:162]: Actually we do because `get_model_args` of the base class is missing the arguments `fb_kwargs`, `fb_type`, `input_type` and `output_type`. If we break backward compatibility, we can rename `fb_type` in `fb_name` and maybe remove `fb_kwargs` but have to keep the other arguments right?
Also, the `get_model_args` of the base class adds `encoder_activation`, `fb_name`, `in_chan` and `out_chan` arguments.

I just created a (very) simple class for the masker of `DeMask` but, unlike all the other maskers, it has to inherit from `nn.Sequential` if we want to be backward compatible. Otherwise, the state dict cannot load. Shall it be moved to `asteroid.masknn`?

Just waiting for comments before adding docstrings and type hinting.

**Comment by jonashaag** [comment]
FWIW I don’t have an opinion on code structure other than that I feel like it is too early to decide on a structure. At this point, to me it feels like Asteroid is mostly a collection of “random” models, with some common building blocks factored out. (Which doesn’t mean that I don’t see the efforts that are going into keeping the code structured and clean, especially by @mpariente who does the vast majority of reviewing. It’s just that the project itself is relatively young, plus there are almost no examples of similar, more mature code bases that you can copy concepts from, simply because the entire business of collecting lots of deep learning models in a single large code base hasn’t been done many times before.) So, unless there is a reason to work on reorganization right now I suggest to wait for the project to grow a little bigger. I’m sure at some point we will have to actively work on reorganization just because we otherwise can’t get any work done anymore, and at that point usually it is much more clear what needs to be done.

**Comment by mpariente** [comment]
In general, I agree that it's too early but I'm trying to be open to suggestions from the main contributors. Anyway, thinking about it isn't useless even if we don't change anything now.

File: asteroid/models/demask.py (hunk 6 above)
Comment (mpariente) [reply asteroid/models/demask.py:162]: About `DeMask`'s masker. Sorry for being hesitant, and thanks for being so adapting. 
We'll have `build_demask_masker` as you write earlier, but as a public method. Inhereting from `Sequential` is a trick for BC, and I don't like it.  

You're (mostly) right about `get_model_args`. Why _mostly_?
- `in_chan` and `out_chan` are not added by `get_model_args`
- Additional kwargs passed to the filterbanks are handled in `Filterbank.get_config()` which makes me see that the current code is wrong wrt to `fb_kwargs` . 

```python
from asteroid import DeMask

model = DeMask(fb_type="param_sinc", min_low_hz=60)
In [53]: model.get_model_args()
Out[53]: 
{'input_type': 'mag',
 'output_type': 'mag',
 'hidden_dims': [1024],
 'dropout': 0.0,
 'activation': 'relu',
 'mask_act': 'relu',
 'norm_type': 'gLN',
 'fb_type': 'param_sinc',
 'n_filters': 512,
 'stride': 256,
 'kernel_size': 512,
 'fb_kwargs': {'min_low_hz': 60},
 'sample_rate': 16000,
 'min_low_hz': 60}
```

Again, thanks for the flexibility. I'll take it from here.

**Comment by mpariente** [comment]
Sorry for force-pushing, but the commit revert didn't work and I didn't spend time to find out why. 

I kept the `fb_type` in the tests on purpose to attest BC. A follow up PR will change to `fb_name`.
